torch_xla/core/dynamo_bridge.py

```diff
@@ -111,6 +111,65 @@ def get_fallback_ops():
   return fallback_ops
 
 
+# Checks that all input args that are tensors are on the same device.
+def _get_input_arg_device(input_args: tuple) -> torch.device:
+  device = None
+  for arg in input_args:
+    if not isinstance(arg, torch.Tensor):
+      continue
+
+    if device is None:
+      device = arg.device
+    else:
+      assert arg.device == device, "Not all args are on the same device."
+
+    return device
+
+
+# Returns True if all the input args are on a CUDA device.
+def _args_on_cuda(input_args: tuple) -> bool:
+  input_device: torch.device = _get_input_arg_device(input_args)
+  if input_device is None:
+    return False
+
+  return input_device.type == "cuda"
+
+
+# Given an input list, moves the tensors to the given target_device.
+# The output order will be the same as the input. Non tensors will also still
+# be in the list.
+def _maybe_move_tensors_to_device(tensors: tuple,
+                                  target_device: torch.device) -> tuple:
+  if not torch.cuda.is_available():
+    return tensors
+
+  moved_tensors = []
+  cpu_device: torch.device = torch.device("cpu")
+
+  for tensor in tensors:
+    if not isinstance(tensor, torch.Tensor):
+      moved_tensors.append(tensor)
+      continue
+
+    if tensor.device == target_device:
+      moved_tensors.append(tensor)
+      continue
+
+    if dynamo_debug:
+      print("Moving Tensor {} to device {}".format(tensor, target_device))
+
+    # Have to move to CPU before moving it to target device.
+    moved_tensor = tensor.to(cpu_device)
+    moved_tensor = moved_tensor.to(target_device)
```

Comment:
I believe you can copy tensors directly to `target_device`.

Reply:
Can we move directly from XLA:GPU to cuda?

Reply:
If I remember correctly, we can, since: https://github.com/pytorch/xla/pull/6060

Reply:
I tried this before and upstream pytorch specifically stated it could not and had to be moved to CPU.

Reply:
This is odd. I'm on 99c41baefa44e62242d158e2321d75b87b01cffe

```python
>>> a = torch.arange(10, device="cuda")
>>> a.to(xm.xla_device())
tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], device='xla:0')
```

Reply:
Oh. Maybe we can't do the other way around.

Reply:
So is this ok to leave here? Thanks for double checking :)